File: pkg/health/workload_rollout.go
package health

import (
	"fmt"
	"strconv"
	"strings"

	appsv1 "k8s.io/api/apps/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
)

type RolloutPhase string

const (
	RolloutIdle             RolloutPhase = "idle"
	RolloutApplying         RolloutPhase = "applying"
	RolloutProgressing      RolloutPhase = "progressing"
	RolloutWaiting          RolloutPhase = "waiting"
	RolloutPaused           RolloutPhase = "paused"
	RolloutPartitionReached RolloutPhase = "partition-reached"
	RolloutStalled          RolloutPhase = "stalled"
)

type WorkloadRolloutActivity struct {
	Phase     RolloutPhase `json:"phase"`
	Active    bool         `json:"active"`
	Manual    bool         `json:"manual"`
	Label     string       `json:"label"`
	Detail    string       `json:"detail,omitempty"`
	Desired   int32        `json:"desired"`
	Updated   int32        `json:"updated"`
	Ready     int32        `json:"ready"`
	Available int32        `json:"available"`
}

func WorkloadRollout(obj any) WorkloadRolloutActivity {
	switch o := obj.(type) {
	case *appsv1.Deployment:
		return deploymentRollout(o)
	case *appsv1.StatefulSet:
		return statefulSetRollout(o)
	case *appsv1.DaemonSet:
		return daemonSetRollout(o)
	case *unstructured.Unstructured:
		if strings.EqualFold(o.GetKind(), "Rollout") && strings.HasPrefix(o.GetAPIVersion(), "argoproj.io/") {
			return argoRollout(o)
		}
	}
	return activity(RolloutIdle, false, "Stable", "", 0, 0, 0, 0)
}

func deploymentRollout(d *appsv1.Deployment) WorkloadRolloutActivity {
	desired := specReplicas(d.Spec.Replicas)
	updated := d.Status.UpdatedReplicas
	ready := d.Status.ReadyReplicas
	available := d.Status.AvailableReplicas
	base := activityCounts(desired, updated, ready, available)

	if d.Status.ObservedGeneration < d.Generation {
		return base.with(RolloutApplying, true, "Applying change", "Waiting for the Deployment controller to observe generation")
	}
	if conditionFailed(d.Status.Conditions, "Progressing", "ProgressDeadlineExceeded") {
		return base.with(RolloutStalled, false, "Rollout stalled", conditionMessage(d.Status.Conditions, "Progressing", "Progress deadline exceeded"))
	}
	old := d.Status.Replicas - updated
	if old < 0 {
		old = 0
	}
	pending := updated < desired || old > 0
	if d.Spec.Paused {
		detail := "No rollout is pending"
		if pending {
			detail = fmt.Sprintf("Paused at %d/%d updated", updated, desired)
		}
		return base.with(RolloutPaused, pending, "Rollout paused", detail)
	}
	if desired == 0 {
		return base.with(RolloutIdle, false, "Scaled to zero", "No replicas requested")
	}
	if !pending {
		return base.with(RolloutIdle, false, "Stable", fmt.Sprintf("%d/%d available", available, desired))
	}
	if d.Spec.Strategy.Type == appsv1.RecreateDeploymentStrategyType {
		if updated == 0 && old > 0 {
			return base.with(RolloutWaiting, true, "Replacing replicas", fmt.Sprintf("Waiting for old replicas to stop · %d available", available))
		}
		if updated == 0 {
			return base.with(RolloutWaiting, true, "Waiting for new revision", replicaDetail(updated, desired, available))
		}
		return base.with(RolloutProgressing, true, "Replacing replicas", replicaDetail(updated, desired, available))
	}
	if updated == 0 {
		return base.with(RolloutWaiting, true, "Waiting for new revision", replicaDetail(updated, desired, available))
	}
	if old == 0 && updated == d.Status.Replicas {
		return base.with(RolloutProgressing, true, "Scaling", replicaDetail(updated, desired, available))
	}
	return base.with(RolloutProgressing, true, "Rolling out", replicaDetail(updated, desired, available))
}

func statefulSetRollout(s *appsv1.StatefulSet) WorkloadRolloutActivity {
	desired := specReplicas(s.Spec.Replicas)
	updated := s.Status.UpdatedReplicas
	ready := s.Status.ReadyReplicas
	base := activityCounts(desired, updated, ready, ready)

	if s.Status.ObservedGeneration < s.Generation {
		return base.with(RolloutApplying, true, "Applying change", "Waiting for the StatefulSet controller to observe generation")
	}
	if desired == 0 {
		return base.with(RolloutIdle, false, "Scaled to zero", "No replicas requested")
	}
	if s.Spec.UpdateStrategy.Type == appsv1.OnDeleteStatefulSetStrategyType {
		if updated == desired && ready == desired {
			return base.with(RolloutIdle, false, "Stable", fmt.Sprintf("%d/%d ready", ready, desired))
		}
		return base.withAction(RolloutWaiting, true, "Waiting for Pod restart", fmt.Sprintf("OnDelete strategy · %d/%d updated", updated, desired))
	}
	partition := int32(0)
	if s.Spec.UpdateStrategy.RollingUpdate != nil && s.Spec.UpdateStrategy.RollingUpdate.Partition != nil {
		partition = *s.Spec.UpdateStrategy.RollingUpdate.Partition
	}
	target := desired - partition
	if target < 0 {
		target = 0
	}
	if updated >= target {
		if partition > 0 && s.Status.CurrentRevision != "" && s.Status.UpdateRevision != "" && s.Status.CurrentRevision != s.Status.UpdateRevision {
			return base.with(RolloutPartitionReached, false, "Partition reached", fmt.Sprintf("%d/%d Pods intentionally retained", partition, desired))
		}
		return base.with(RolloutIdle, false, "Stable", fmt.Sprintf("%d/%d ready", ready, desired))
	}
	if updated == 0 {
		return base.with(RolloutWaiting, true, "Waiting for new revision", fmt.Sprintf("%d/%d target Pods updated · %d/%d ready", updated, target, ready, desired))
	}
	return base.with(RolloutProgressing, true, "Rolling out", fmt.Sprintf("%d/%d target Pods updated · %d/%d ready", updated, target, ready, desired))
}

func daemonSetRollout(d *appsv1.DaemonSet) WorkloadRolloutActivity {
	desired := d.Status.DesiredNumberScheduled
	updated := d.Status.UpdatedNumberScheduled
	ready := d.Status.NumberReady
	available := d.Status.NumberAvailable
	base := activityCounts(desired, updated, ready, available)

	if d.Status.ObservedGeneration < d.Generation {
		return base.with(RolloutApplying, true, "Applying change", "Waiting for the DaemonSet controller to observe generation")
	}
	if desired == 0 {
		return base.with(RolloutIdle, false, "No targets", "Selector matches no nodes")
	}
	if d.Spec.UpdateStrategy.Type == appsv1.OnDeleteDaemonSetStrategyType {
		if updated == desired && available == desired {
			return base.with(RolloutIdle, false, "Stable", fmt.Sprintf("%d/%d available", available, desired))
		}
		return base.withAction(RolloutWaiting, true, "Waiting for Pod restart", fmt.Sprintf("OnDelete strategy · %d/%d updated", updated, desired))
	}
	if updated == desired {
		return base.with(RolloutIdle, false, "Stable", fmt.Sprintf("%d/%d available", available, desired))
	}
	if updated == 0 {
		return base.with(RolloutWaiting, true, "Waiting for new revision", replicaDetail(updated, desired, available))
	}
	return base.with(RolloutProgressing, true, "Rolling out", replicaDetail(updated, desired, available))
}

func argoRollout(r *unstructured.Unstructured) WorkloadRolloutActivity {
	desired := nestedInt32(r.Object, "spec", "replicas")
	if _, found, _ := unstructured.NestedFieldNoCopy(r.Object, "spec", "replicas"); !found {
		desired = nestedInt32(r.Object, "status", "replicas")
		if desired == 0 {
			desired = 1
		}
	}
	updated := nestedInt32(r.Object, "status", "updatedReplicas")
	ready := nestedInt32(r.Object, "status", "readyReplicas")
	available := nestedInt32(r.Object, "status", "availableReplicas")
	base := activityCounts(desired, updated, ready, available)
	phase, _, _ := unstructured.NestedString(r.Object, "status", "phase")
	message, _, _ := unstructured.NestedString(r.Object, "status", "message")
	aborted, _, _ := unstructured.NestedBool(r.Object, "status", "abort")
	if observed, found, _ := unstructured.NestedString(r.Object, "status", "observedGeneration"); found {
		generation, err := strconv.ParseInt(observed, 10, 64)
		if err == nil && generation > 0 && generation < r.GetGeneration() {
			return base.with(RolloutApplying, true, "Applying change", "Waiting for the Rollout controller to observe generation")
		}
	}
	if failedMessage, failed := argoFailureMessage(r); failed {
		if failedMessage == "" {
			failedMessage = message
		}
		if failedMessage == "" {
			failedMessage = "The Rollout controller reported a failed revision"
		}
		return base.with(RolloutStalled, false, "Rollout stalled", failedMessage)
	}
	if aborted {
		if message == "" {
			message = "The Rollout was aborted"
		}
		return base.with(RolloutStalled, false, "Rollout stalled", message)
	}
	switch strings.ToLower(phase) {
	case "degraded", "error", "failed", "aborted":
		if message == "" {
			message = "The Rollout controller reported a failed revision"
		}
		return base.with(RolloutStalled, false, "Rollout stalled", message)
	case "paused":
		return base.with(RolloutPaused, true, "Rollout paused", argoStepDetail(r, updated, desired, available))
	case "progressing":
		if updated == 0 {
			return base.with(RolloutWaiting, true, "Waiting for new revision", argoStepDetail(r, updated, desired, available))
		}
		return base.with(RolloutProgressing, true, "Rolling out", argoStepDetail(r, updated, desired, available))
	case "healthy":
		return base.with(RolloutIdle, false, "Stable", fmt.Sprintf("%d/%d available", available, desired))
	default:
		if desired > 0 && (updated < desired || available < desired) {
			return base.with(RolloutApplying, true, "Applying change", argoStepDetail(r, updated, desired, available))
		}
		return base.with(RolloutIdle, false, "Stable", message)
	}
}

func argoFailureMessage(r *unstructured.Unstructured) (string, bool) {
	conditions, _, _ := unstructured.NestedSlice(r.Object, "status", "conditions")
	for _, raw := range conditions {
		condition, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		conditionType, _ := condition["type"].(string)
		conditionStatus, _ := condition["status"].(string)
		reason, _ := condition["reason"].(string)
		failed := (conditionType == "InvalidSpec" && conditionStatus == "True") ||
			(conditionType == "Progressing" && (conditionStatus == "False" || reason == "ProgressDeadlineExceeded"))
		if !failed {
			continue
		}
		if message, _ := condition["message"].(string); message != "" {
			return message, true
		}
		return reason, true
	}
	return "", false
}

func argoStepDetail(r *unstructured.Unstructured, updated, desired, available int32) string {
	steps, stepsFound, _ := unstructured.NestedSlice(r.Object, "spec", "strategy", "canary", "steps")
	step, found, _ := unstructured.NestedInt64(r.Object, "status", "currentStepIndex")
	if !found || !stepsFound || len(steps) == 0 {
		return replicaDetail(updated, desired, available)
	}
	displayStep := step + 1
	if displayStep < 1 {
		displayStep = 1
	}
	if displayStep > int64(len(steps)) {
		displayStep = int64(len(steps))
	}
	label := ""
	if currentStep, ok := steps[displayStep-1].(map[string]any); ok {
		label = fmt.Sprintf(" (%s)", canaryStepLabel(currentStep))
	}
	weightSuffix := ""
	if weight, wFound, _ := unstructured.NestedInt64(r.Object, "status", "canary", "weights", "canary", "weight"); wFound {
		weightSuffix = fmt.Sprintf(" · %d%% canary traffic", weight)
	}
	return fmt.Sprintf("Step %d%s · %d/%d updated · %d available%s", displayStep, label, updated, desired, available, weightSuffix)
}

// canaryStepLabel mirrors packages/k8s-ui/src/components/resources/renderers/
// RolloutRenderer.tsx's canaryStepLabel exactly (same output strings, same
// step-type coverage) - the two are checked against the same golden fixture
// (testdata/workload_rollout_vectors.json), so a change to one without the
// other silently breaks cross-language parity rather than failing loudly.
func canaryStepLabel(step map[string]any) string {
	if len(step) == 0 {
		return "Unknown step"
	}

	if weight, ok := step["setWeight"]; ok {
		return fmt.Sprintf("Set weight: %s%%", stepNumberString(weight))
	}

	if pause, ok := step["pause"].(map[string]any); ok {
		if duration, ok := pause["duration"]; ok && duration != nil && duration != "" {
			return fmt.Sprintf("Pause: %v", duration)
		}
		return "Pause: until promoted"
	}
	if _, ok := step["pause"]; ok {
		return "Pause: until promoted"
	}

	if analysis, ok := step["analysis"].(map[string]any); ok {
		names := templateNames(analysis["templates"])
		if len(names) > 0 {
			return fmt.Sprintf("Analysis: %s", strings.Join(names, ", "))
		}
		return "Analysis"
	}

	if experiment, ok := step["experiment"].(map[string]any); ok {
		names := experimentTemplateNames(experiment["templates"])
		duration := ""
		if d, ok := experiment["duration"]; ok && d != nil && d != "" {
			duration = fmt.Sprintf(" for %v", d)
		}
		if len(names) > 0 {
			return fmt.Sprintf("Experiment: %s%s", strings.Join(names, ", "), duration)
		}
		return fmt.Sprintf("Experiment%s", duration)
	}

	if scale, ok := step["setCanaryScale"].(map[string]any); ok {
		if match, _ := scale["matchTrafficWeight"].(bool); match {
			return "Set canary scale: match traffic weight"
		}
		if replicas, ok := scale["replicas"]; ok {
			return fmt.Sprintf("Set canary scale: %s replicas", stepNumberString(replicas))
		}
		if weight, ok := scale["weight"]; ok {
			return fmt.Sprintf("Set canary scale: %s%%", stepNumberString(weight))
		}
		return "Set canary scale"
	}

	if route, ok := step["setHeaderRoute"].(map[string]any); ok {
		name, _ := route["name"].(string)
		match, _ := route["match"].([]any)
		if len(match) == 0 {
			if name != "" {
				return fmt.Sprintf("Remove header route: %s", name)
			}
			return "Remove header route"
		}
		var headers []string
		for _, m := range match {
			if mm, ok := m.(map[string]any); ok {
				if headerName, ok := mm["headerName"].(string); ok && headerName != "" {
					headers = append(headers, headerName)
				}
			}
		}
		label := "Header route"
		if name != "" {
			label += " " + name
		}
		if len(headers) > 0 {
			label += ": " + strings.Join(headers, ", ")
		}
		return label
	}

	if route, ok := step["setMirrorRoute"].(map[string]any); ok {
		name, _ := route["name"].(string)
		match, _ := route["match"].([]any)
		if len(match) == 0 {
			if name != "" {
				return fmt.Sprintf("Remove mirror route: %s", name)
			}
			return "Remove mirror route"
		}
		label := "Mirror route"
		if name != "" {
			label += " " + name
		}
		if pct, ok := route["percentage"]; ok {
			label += fmt.Sprintf(" (%s%%)", stepNumberString(pct))
		}
		return label
	}

	if plugin, ok := step["plugin"].(map[string]any); ok {
		name, _ := plugin["name"].(string)
		if name == "" {
			name = "unnamed"
		}
		return fmt.Sprintf("Plugin: %s", name)
	}

	for key := range step {
		return fmt.Sprintf("Unrecognized step: %s", key)
	}
	return "Unknown step"
}

// templateNames extracts templateName/clusterTemplateName from an
// analysis.templates list (AnalysisTemplate/ClusterAnalysisTemplate refs),
// same as the TS version. NOT used for experiment.templates — those are pod
// templates, a different shape entirely; see experimentTemplateNames.
func templateNames(raw any) []string {
	list, ok := raw.([]any)
	if !ok {
		return nil
	}
	var names []string
	for _, item := range list {
		t, ok := item.(map[string]any)
		if !ok {
			continue
		}
		if name, ok := t["templateName"].(string); ok && name != "" {
			names = append(names, name)
			continue
		}
		if name, ok := t["clusterTemplateName"].(string); ok && name != "" {
			names = append(names, name)
		}
	}
	return names
}

// experimentTemplateNames extracts `name` from an experiment.templates
// list — Argo's RolloutExperimentTemplate entries (pod template variants
// like "baseline"/"canary" an Experiment spins up replicas of) carry their
// identifier under `name`, not `templateName`/`clusterTemplateName` (the
// AnalysisTemplate ref shape templateNames handles). Mirrors the TS version.
func experimentTemplateNames(raw any) []string {
	list, ok := raw.([]any)
	if !ok {
		return nil
	}
	var names []string
	for _, item := range list {
		t, ok := item.(map[string]any)
		if !ok {
			continue
		}
		if name, ok := t["name"].(string); ok && name != "" {
			names = append(names, name)
		}
	}
	return names
}

// stepNumberString formats a JSON-decoded numeric value the way it appears
// in the Rollout spec (e.g. int64(20) -> "20"), matching JS's implicit
// number->string coercion used by the TS version's template literals.
func stepNumberString(v any) string {
	switch n := v.(type) {
	case int64:
		return strconv.FormatInt(n, 10)
	case float64:
		return strconv.FormatFloat(n, 'f', -1, 64)
	default:
		return fmt.Sprintf("%v", v)
	}
}

func conditionFailed(conditions []appsv1.DeploymentCondition, conditionType appsv1.DeploymentConditionType, reason string) bool {
	for _, condition := range conditions {
		if condition.Type == conditionType && (condition.Status == "False" || condition.Reason == reason) {
			return true
		}
	}
	return false
}

func conditionMessage(conditions []appsv1.DeploymentCondition, conditionType appsv1.DeploymentConditionType, fallback string) string {
	for _, condition := range conditions {
		if condition.Type == conditionType && condition.Message != "" {
			return condition.Message
		}
	}
	return fallback
}

func nestedInt32(obj map[string]any, fields ...string) int32 {
	value, found, _ := unstructured.NestedInt64(obj, fields...)
	if found {
		return int32(value)
	}
	return 0
}

func activityCounts(desired, updated, ready, available int32) WorkloadRolloutActivity {
	return activity(RolloutIdle, false, "Stable", "", desired, updated, ready, available)
}

func activity(phase RolloutPhase, active bool, label, detail string, desired, updated, ready, available int32) WorkloadRolloutActivity {
	return WorkloadRolloutActivity{Phase: phase, Active: active, Label: label, Detail: detail, Desired: desired, Updated: updated, Ready: ready, Available: available}
}

func (a WorkloadRolloutActivity) with(phase RolloutPhase, active bool, label, detail string) WorkloadRolloutActivity {
	a.Phase = phase
	a.Active = active
	a.Manual = false
	a.Label = label
	a.Detail = detail
	return a
}

func (a WorkloadRolloutActivity) withAction(phase RolloutPhase, active bool, label, detail string) WorkloadRolloutActivity {
	a = a.with(phase, active, label, detail)
	a.Manual = true
	return a
}

func replicaDetail(updated, desired, available int32) string {
	return fmt.Sprintf("%d/%d updated · %d available", updated, desired, available)
}

File: packages/k8s-ui/src/utils/workload-rollout.ts
import type { WorkloadPodInfo } from '../types/core'
import { canaryStepLabel } from '../components/resources/renderers/RolloutRenderer'

export type WorkloadRolloutPhase =
  | 'idle'
  | 'applying'
  | 'progressing'
  | 'waiting'
  | 'paused'
  | 'partition-reached'
  | 'stalled'

export interface WorkloadRolloutActivity {
  phase: WorkloadRolloutPhase
  active: boolean
  manual: boolean
  label: string
  detail?: string
  desired: number
  updated: number
  ready: number
  available: number
}

export function isArgoRolloutResource(resource: any): boolean {
  return String(resource?.apiVersion || '').startsWith('argoproj.io/')
}

export function rolloutMayAdvanceAutomatically(activity: WorkloadRolloutActivity): boolean {
  return activity.phase === 'applying' ||
    activity.phase === 'progressing' ||
    (activity.phase === 'waiting' && !activity.manual)
}

export function isRolloutActivityVisible(activity: WorkloadRolloutActivity): boolean {
  return activity.active || activity.phase === 'stalled'
}

export function getObservedGeneration(resource: any): number {
  const observed = resource?.status?.observedGeneration
  if (typeof observed === 'number' && Number.isFinite(observed)) return observed
  return typeof observed === 'string' && /^\d+$/.test(observed)
    ? Number.parseInt(observed, 10)
    : 0
}

export function getWorkloadRolloutActivity(
  resource: any,
  kind?: string,
  workloadPods?: WorkloadPodInfo[],
): WorkloadRolloutActivity {
  const normalized = (kind || resource?.kind || '').toLowerCase().replace(/s$/, '')
  if (normalized === 'rollout' && !isArgoRolloutResource(resource)) {
    return activity('idle', false, false, 'Stable', '', 0, 0, 0, 0)
  }
  let result: WorkloadRolloutActivity
  switch (normalized) {
    case 'deployment':
      result = deploymentActivity(resource)
      break
    case 'statefulset':
      result = statefulSetActivity(resource)
      break
    case 'daemonset':
      result = daemonSetActivity(resource)
      break
    case 'rollout':
      result = argoRolloutActivity(resource)
      break
    default:
      result = activity('idle', false, false, 'Stable', '', 0, 0, 0, 0)
  }
  return withUpdatedRevisionFailure(result, workloadPods)
}

function deploymentActivity(resource: any): WorkloadRolloutActivity {
  const spec = resource?.spec || {}
  const status = resource?.status || {}
  const desired = spec.replicas ?? 1
  const updated = status.updatedReplicas ?? 0
  const ready = status.readyReplicas ?? 0
  const available = status.availableReplicas ?? 0
  const base = activity('idle', false, false, 'Stable', '', desired, updated, ready, available)
  const progress = (status.conditions || []).find((condition: any) => condition.type === 'Progressing')

  if ((status.observedGeneration ?? 0) < (resource?.metadata?.generation ?? 0)) {
    return merge(base, 'applying', true, 'Applying change', 'Waiting for the Deployment controller to observe generation')
  }
  if (progress?.status === 'False' || progress?.reason === 'ProgressDeadlineExceeded') {
    return merge(base, 'stalled', false, 'Rollout stalled', progress.message || 'Progress deadline exceeded')
  }
  const old = Math.max(0, (status.replicas ?? 0) - updated)
  const pending = updated < desired || old > 0
  if (spec.paused) {
    return merge(base, 'paused', pending, 'Rollout paused', pending ? `Paused at ${updated}/${desired} updated` : 'No rollout is pending')
  }
  if (desired === 0) return merge(base, 'idle', false, 'Scaled to zero', 'No replicas requested')
  if (!pending) return merge(base, 'idle', false, 'Stable', `${available}/${desired} available`)
  if (spec.strategy?.type === 'Recreate') {
    if (updated === 0 && old > 0) {
      return merge(base, 'waiting', true, 'Replacing replicas', `Waiting for old replicas to stop · ${available} available`)
    }
    return updated === 0
      ? merge(base, 'waiting', true, 'Waiting for new revision', replicaDetail(updated, desired, available))
      : merge(base, 'progressing', true, 'Replacing replicas', replicaDetail(updated, desired, available))
  }
  if (updated === 0) return merge(base, 'waiting', true, 'Waiting for new revision', replicaDetail(updated, desired, available))
  return old === 0 && updated === (status.replicas ?? 0)
    ? merge(base, 'progressing', true, 'Scaling', replicaDetail(updated, desired, available))
    : merge(base, 'progressing', true, 'Rolling out', replicaDetail(updated, desired, available))
}

function statefulSetActivity(resource: any): WorkloadRolloutActivity {
  const spec = resource?.spec || {}
  const status = resource?.status || {}
  const desired = spec.replicas ?? 1
  const updated = status.updatedReplicas ?? 0
  const ready = status.readyReplicas ?? 0
  const base = activity('idle', false, false, 'Stable', '', desired, updated, ready, ready)

  if ((status.observedGeneration ?? 0) < (resource?.metadata?.generation ?? 0)) {
    return merge(base, 'applying', true, 'Applying change', 'Waiting for the StatefulSet controller to observe generation')
  }
  if (desired === 0) return merge(base, 'idle', false, 'Scaled to zero', 'No replicas requested')
  if (spec.updateStrategy?.type === 'OnDelete') {
    return updated === desired && ready === desired
      ? merge(base, 'idle', false, 'Stable', `${ready}/${desired} ready`)
      : mergeManual(base, 'waiting', true, 'Waiting for Pod restart', `OnDelete strategy · ${updated}/${desired} updated`)
  }
  const partition = spec.updateStrategy?.rollingUpdate?.partition ?? 0
  const target = Math.max(0, desired - partition)
  if (updated >= target) {
    const pendingRevision = Boolean(status.currentRevision && status.updateRevision && status.currentRevision !== status.updateRevision)
    return partition > 0 && pendingRevision
      ? merge(base, 'partition-reached', false, 'Partition reached', `${partition}/${desired} Pods intentionally retained`)
      : merge(base, 'idle', false, 'Stable', `${ready}/${desired} ready`)
  }
  const detail = `${updated}/${target} target Pods updated · ${ready}/${desired} ready`
  return updated === 0
    ? merge(base, 'waiting', true, 'Waiting for new revision', detail)
    : merge(base, 'progressing', true, 'Rolling out', detail)
}

function daemonSetActivity(resource: any): WorkloadRolloutActivity {
  const spec = resource?.spec || {}
  const status = resource?.status || {}
  const desired = status.desiredNumberScheduled ?? 0
  const updated = status.updatedNumberScheduled ?? 0
  const ready = status.numberReady ?? 0
  const available = status.numberAvailable ?? 0
  const base = activity('idle', false, false, 'Stable', '', desired, updated, ready, available)

  if ((status.observedGeneration ?? 0) < (resource?.metadata?.generation ?? 0)) {
    return merge(base, 'applying', true, 'Applying change', 'Waiting for the DaemonSet controller to observe generation')
  }
  if (desired === 0) return merge(base, 'idle', false, 'No targets', 'Selector matches no nodes')
  if (spec.updateStrategy?.type === 'OnDelete') {
    return updated === desired && available === desired
      ? merge(base, 'idle', false, 'Stable', `${available}/${desired} available`)
      : mergeManual(base, 'waiting', true, 'Waiting for Pod restart', `OnDelete strategy · ${updated}/${desired} updated`)
  }
  if (updated === desired) return merge(base, 'idle', false, 'Stable', `${available}/${desired} available`)
  return updated === 0
    ? merge(base, 'waiting', true, 'Waiting for new revision', replicaDetail(updated, desired, available))
    : merge(base, 'progressing', true, 'Rolling out', replicaDetail(updated, desired, available))
}

function argoRolloutActivity(resource: any): WorkloadRolloutActivity {
  const spec = resource?.spec || {}
  const status = resource?.status || {}
  const desired = spec.replicas ?? ((status.replicas ?? 0) || 1)
  const updated = status.updatedReplicas ?? 0
  const ready = status.readyReplicas ?? 0
  const available = status.availableReplicas ?? 0
  const base = activity('idle', false, false, 'Stable', '', desired, updated, ready, available)
  const detail = argoDetail(resource, updated, desired, available)
  const observedGeneration = getObservedGeneration(resource)
  if (observedGeneration > 0 && observedGeneration < (resource?.metadata?.generation ?? 0)) {
    return merge(base, 'applying', true, 'Applying change', 'Waiting for the Rollout controller to observe generation')
  }
  const failedCondition = (status.conditions || []).find((condition: any) =>
    (condition.type === 'InvalidSpec' && condition.status === 'True') ||
    (condition.type === 'Progressing' && (condition.status === 'False' || condition.reason === 'ProgressDeadlineExceeded')),
  )
  if (failedCondition || status.abort) {
    return merge(base, 'stalled', false, 'Rollout stalled', failedCondition?.message || failedCondition?.reason || status.message || (status.abort ? 'The Rollout was aborted' : '') || 'The Rollout controller reported a failed revision')
  }
  switch (String(status.phase || '').toLowerCase()) {
    case 'degraded':
    case 'error':
    case 'failed':
    case 'aborted':
      return merge(base, 'stalled', false, 'Rollout stalled', status.message || 'The Rollout controller reported a failed revision')
    case 'paused':
      return merge(base, 'paused', true, 'Rollout paused', detail)
    case 'progressing':
      return updated === 0
        ? merge(base, 'waiting', true, 'Waiting for new revision', detail)
        : merge(base, 'progressing', true, 'Rolling out', detail)
    case 'healthy':
      return merge(base, 'idle', false, 'Stable', `${available}/${desired} available`)
    default:
      return desired > 0 && (updated < desired || available < desired)
        ? merge(base, 'applying', true, 'Applying change', detail)
        : merge(base, 'idle', false, 'Stable', status.message || '')
  }
}

function withUpdatedRevisionFailure(
  rollout: WorkloadRolloutActivity,
  pods?: WorkloadPodInfo[],
): WorkloadRolloutActivity {
  if ((!rollout.active && rollout.phase !== 'stalled') || rollout.phase === 'applying' || !pods) return rollout
  const failed = pods
    .filter((pod) => pod.updatedRevision === true && pod.healthLevel === 'unhealthy')
    .sort((a, b) => a.name.localeCompare(b.name))[0]
  if (!failed) return rollout
  const reason = failed.reason || failed.lastTerminationReason || failed.phase || 'Pod failed'
  return merge(rollout, 'stalled', false, 'New revision cannot start', `${reason} · Pod ${failed.name}`)
}

export function getArgoRolloutStepNumber(resource: any): number | null {
  const steps = resource?.spec?.strategy?.canary?.steps || []
  const currentIndex = resource?.status?.currentStepIndex
  if (steps.length === 0 || typeof currentIndex !== 'number') return null
  return Math.min(Math.max(currentIndex + 1, 1), steps.length)
}

// Mirrors pkg/health/workload_rollout.go's argoStepDetail exactly (same
// output string, same step-type coverage via canaryStepLabel) - the two are
// checked against the same golden fixture
// (pkg/health/testdata/workload_rollout_vectors.json), so a change to one
// without the other silently breaks cross-language parity rather than
// failing loudly.
function argoDetail(resource: any, updated: number, desired: number, available: number): string {
  const steps = resource?.spec?.strategy?.canary?.steps || []
  const stepNumber = getArgoRolloutStepNumber(resource)
  const step = stepNumber === null ? '' : `Step ${stepNumber}`
  const currentStep = stepNumber === null ? undefined : steps[stepNumber - 1]
  const label = currentStep ? ` (${canaryStepLabel(currentStep)})` : ''
  const weight = resource?.status?.canary?.weights?.canary?.weight
  const weightSuffix = typeof weight === 'number' ? ` · ${weight}% canary traffic` : ''
  const prefix = stepNumber === null ? '' : `${step}${label} · `
  return `${prefix}${updated}/${desired} updated · ${available} available${weightSuffix}`
}

function replicaDetail(updated: number, desired: number, available: number): string {
  return `${updated}/${desired} updated · ${available} available`
}

function activity(
  phase: WorkloadRolloutPhase,
  active: boolean,
  manual: boolean,
  label: string,
  detail: string,
  desired: number,
  updated: number,
  ready: number,
  available: number,
): WorkloadRolloutActivity {
  return { phase, active, manual, label, detail, desired, updated, ready, available }
}

function merge(
  base: WorkloadRolloutActivity,
  phase: WorkloadRolloutPhase,
  active: boolean,
  label: string,
  detail: string,
): WorkloadRolloutActivity {
  return { ...base, phase, active, manual: false, label, detail }
}

function mergeManual(
  base: WorkloadRolloutActivity,
  phase: WorkloadRolloutPhase,
  active: boolean,
  label: string,
  detail: string,
): WorkloadRolloutActivity {
  return { ...base, phase, active, manual: true, label, detail }
}

export function rolloutActivityLevel(activity: WorkloadRolloutActivity): 'healthy' | 'neutral' | 'degraded' | 'unhealthy' {
  switch (activity.phase) {
    case 'stalled':
      return 'unhealthy'
    case 'paused':
      return 'degraded'
    case 'waiting':
      return activity.manual ? 'degraded' : 'neutral'
    case 'applying':
    case 'progressing':
    case 'partition-reached':
      return 'neutral'
    default:
      return activity.label === 'Stable' ? 'healthy' : 'neutral'
  }
}

export function rolloutActivityBadge(activity: WorkloadRolloutActivity): { text: string; color: string; level: 'healthy' | 'neutral' | 'degraded' | 'unhealthy' } {
  const level = rolloutActivityLevel(activity)
  return { text: activity.label, color: `status-${level}`, level }
}
